### PR TITLE
osd/PeeringState: add perf counters for PG rebuild times

### PR DESCRIPTION
**Pull request by sseshasa**
Track per-OSD PG rebuild duration in prepare_stats_for_publish() (primary OSD only). Three new counters are added to the recoverystate_perf collection:

 - `pg_rebuild_duration`: LONGRUNAVG time counter (sum+count pair); This counter internally maintains 'avgcount' that tracks the cumulative number of rebuild events.
 - `pg_rebuild_max_secs`: maximum rebuild duration observed in seconds
 - `pg_rebuild_min_secs`: minimum rebuild duration observed in seconds

The logic uses a per-PG in-memory latch (`rebuild_start_time`) to capture the redundancy failure entry point. When a PG is first observed to be in a vulnerable state (PG_STATE_DEGRADED,
PG_STATE_UNDERSIZED, or num_objects_misplaced/degraded > 0), it latches info.stats.last_change as the start time, provided last_change > last_clean, which ensures only genuine new failures after the prior clean interval are tracked.

On recovery, the rebuild duration is computed as (`now - rebuild_start_time`) and is only recorded if delta `num_objects_recovered > 0` or the PG had confirmed redundancy loss at latch time, filtering out spurious state transitions. The latch is cleared after each recorded event.

The latch is also cleared in `clear_primary_state()` so that an interval change or role transition (primary -> replica) does not carry a stale start time or baseline recovered count into a future interval.

The new last_degraded is intentionally not used here to retain compatibility with older Ceph branches where the field doesn't exist and requires encoding changes. A future simplification can replace the latch with a direct (`last_clean - last_degraded`) calculation once last_degraded is consistently available.

This interim solution is a close approximation of the PG rebuild time.

These counters are scraped per-OSD by ceph-exporter and exposed to Prometheus, enabling durability score calculations over user-defined time windows.

Other Changes:
1. Add unit tests to TestPeeringState.cc that exercise the latch logic.
2. Add a standalone integration test to verify that the rebuild perf counters are incremented on the primary OSD after a recovery event.

Fixes: https://tracker.ceph.com/issues/77493
Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

### CODE REVIEW

**Comment by sseshasa** [comment]
Here's an example of how the stats look like:

```
"recoverystate_perf": {
        ...
        "process_log_stats_invalidated": 0,
        "pg_split_parent_stats_invalidated": 0,
        "pg_split_child_stats_invalidated": 0,
        "update_stats_invalidated": 0,
        "append_log_stats_invalidated": 0,
        "merge_log_stats_invalidated": 0,
        "pg_rebuild_duration": {
            "avgcount": 1,
            "sum": 0.341132220,
            "avgtime": 0.341132220
        },
        "pg_rebuild_max_secs": 0,
        "pg_rebuild_min_secs": 0
    },

```

File: src/osd/osd_perf_counters.h [hunk 1]
```diff
@@ -263,6 +263,9 @@ enum {
   rs_update_stats_invalidated,
   rs_append_log_stats_invalidated,
   rs_merge_log_stats_invalidated,
+  rs_pg_rebuild_latency,
```
Comment (ronen-fr) [inline src/osd/osd_perf_counters.h]: I'm not sure 'latency' is the best term here. As the '.cc' comment says - you seem to be measuring 'duration' of a process, not 'latency' delay until it starts.

File: src/osd/PeeringState.h [hunk 2]
```diff
@@ -1579,6 +1579,14 @@ class PeeringState : public MissingLoc::MappingInfo {
   bool backfill_reserved = false;
   bool backfill_reserving = false;
 
+  /**
+   * Per-PG latch state for rebuild time tracking. Cleared after each
+   * completed rebuild event is recorded in the perf counters.
+   */
+  utime_t rebuild_start_time;
+  int64_t rebuild_base_recovered = 0;
+  bool rebuild_had_redundancy_loss = false;
```
Comment (ronen-fr) [inline src/osd/PeeringState.h:1591]: is this also cleared on 'unusual' state changes (interval etc')?

**Review by ronen-fr — APPROVED** [review]
LGTM with two comments.

File: src/osd/osd_perf_counters.h (hunk 1 above)
Comment (sseshasa) [reply src/osd/osd_perf_counters.h]: That's correct. I initially added this as "latency" but didn't change it eventually. I changed it everywhere now. Please verify.

File: src/osd/PeeringState.h (hunk 2 above)
Comment (sseshasa) [reply src/osd/PeeringState.h:1591]: Thanks for pointing this out. This was in fact not being done.

I added code to reset them in `clear_primary_state()` so that an interval change or role transition
(primary -> replica) during recovery does not carry a stale  start time or baseline recovered count into
a future interval.

I also added a unit test to ensure that a role change during recovery ensures that the above variables are cleared.

I audited other PG states to confirm if they warrant clearing and confirmed that resetting them in `clear_primary_state()` is the right place.

**Comment by ronen-fr** [comment]
jenkins test api

**Comment by sseshasa** [comment]
jenkins test make check

**Comment by sseshasa** [comment]
jenkins test make check

**Comment by ljflores** [comment]
make check failed from https://tracker.ceph.com/issues/77589

api check failed from https://tracker.ceph.com/issues/77618

**Comment by sseshasa** [comment]
jenkins test make check

**Comment by MattyWilliams22** [comment]
jenkins test api

**Comment by NitzanMordhai** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues77852
